File: src/main/java/io/github/emcw/map/ProcessedTown.java
package io.github.emcw.map;

import com.google.gson.JsonObject;
import org.jetbrains.annotations.NotNull;
import org.jsoup.Jsoup;
import org.jsoup.safety.Safelist;

import java.util.List;
import java.util.stream.Collectors;

import static io.github.emcw.utils.Funcs.calcArea;
import static io.github.emcw.utils.GsonUtil.*;

@SuppressWarnings("unused")
public class ProcessedTown {
    public final String name, desc;
    public final String fill, outline;

    public final int[] x, z;
    public final int area;

    public final List<String> info;

    static final Safelist whitelist = new Safelist().addAttributes("a", "href");

    public ProcessedTown(JsonObject obj) {
        this.name = keyAsStr(obj, "label");
        this.desc = keyAsStr(obj, "desc");

        this.info = this.desc != null ? processFlags(this.desc) : null;

        this.fill = keyAsStr(obj, "fillcolor");
        this.outline = keyAsStr(obj, "color");

        this.x = arrToIntArr(keyAsArr(obj, "x"));
        this.z = arrToIntArr(keyAsArr(obj, "z"));
        this.area = calcArea(this.x, this.z);
    }

    public static List<String> processFlags(@NotNull String str) {
        return strArrAsStream(str.split("<br />"))
                .map(e -> Jsoup.clean(e, whitelist))
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/io/github/emcw/map/entities/ProcessedTown.kt
package io.github.emcw.map.entities

import com.google.gson.JsonObject
import io.github.emcw.utils.Funcs
import io.github.emcw.utils.GsonUtil
import org.jsoup.Jsoup
import org.jsoup.safety.Safelist
import java.util.stream.Collectors

@Suppress("unused")
class ProcessedTown(obj: JsonObject) {
    val name: String?
    val desc: String?
    val fill: String?
    val outline: String?
    val x: IntArray?
    val z: IntArray?
    val area: Int
    val info: List<String>?

    init {
        name = GsonUtil.keyAsStr(obj, "label")
        desc = GsonUtil.keyAsStr(obj, "desc")
        info = if (desc != null) processFlags(desc) else null
        fill = GsonUtil.keyAsStr(obj, "fillcolor")
        outline = GsonUtil.keyAsStr(obj, "color")
        x = GsonUtil.arrToIntArr(GsonUtil.keyAsArr(obj, "x"))
        z = GsonUtil.arrToIntArr(GsonUtil.keyAsArr(obj, "z"))
        area = Funcs.calcArea(x, z)
    }

    companion object {
        val whitelist = Safelist().addAttributes("a", "href")
        fun processFlags(str: String): List<String> {
            return GsonUtil.strArrAsStream(str.split("<br />".toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray())
                .map { e: String? -> Jsoup.clean(e, whitelist) }
                .collect(Collectors.toList())
        }
    }
}